chore(static-site): hash uploaded content for asset mapping file

dannyrandall opened the pull request:
This change makes it so that the **asset mapping file name** is now a **hash of all the uploaded content**. In the template package (see https://github.com/aws/copilot-cli/pull/4762), the path to that file is passed to the custom resource that triggers the copy state machine. By making the file name based on a hash of the content, the state machine will only be triggered if any of the uploaded content changes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

dannyrandall committed:
make asset mapping path be a hash of uploaded content

dannyrandall committed:
update struct fields, return asset mapping path

dannyrandall committed:
better struct field

dannyrandall committed:
undo gen mocks on `mock_task.go`

Lou1415926 commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -29,60 +30,62 @@ type ArtifactBucketUploader struct {
 }
 
 // UploadFiles hashes each of the files specified in files and uploads
 // them to the path "{PathPrefix}/{hash}". After, it uploads a JSON file
```
```suggestion
// them to the path "{AssetDir}/{hash}". After, it uploads a JSON file
```

Lou1415926 commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -13,6 +13,7 @@ import (
 	"fmt"
 	"io"
 	"io/fs"
+	"path"
```
ooohhh good catch!

Lou1415926 commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -140,30 +143,40 @@ func (u *ArtifactBucketUploader) uploadAssets(assets []asset) error {
+//
+// The uploaded path of the file is returned along with an error, if any.
+func (u *ArtifactBucketUploader) uploadAssetMappingFile(assets []asset) (string, error) {
 	sort.Slice(assets, func(i, j int) bool {
```
Will this ever result in a equal comparison (I assume no - it probably would never make sense for two files to be uploaded to the same bucket and the same path, no matter how this package evolves 🤔 💭  ..)? If so, just in case, maybe we want to use https://pkg.go.dev/sort#SliceStable

dannyrandall replied:
it _is_ possible that two mappings are the exact same - though likely not what the user is intending to configure. e.g., two uploads: one of folder `./static` and one of file `./static/index.html` with `dest: ./static` would result in _two_ `index.html hash` -> `/static/index.html` entries in assets

i think the right approach here is probably just to remove any duplicate elements instead of using slicestable - thoughts?

Lou1415926 replied:
> e.g., two uploads: one of folder ./static and one of file ./static/index.html with dest: ./static would result in two index.html hash -> /static/index.html entries in assets

Makes sense! In this case, the user could have just not specified `./static/index.html` right? If we don't remove the duplicated entries, it wouldn't have caused issues because we basically just upload the same thing twice. If this is the case, then

> i think the right approach here is probably just to remove any duplicate elements instead of using slicestable


Agreed!

dannyrandall replied:
> the user could have just not specified ./static/index.html right?

yeah, exactly! 👍

Lou1415926 commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -140,30 +143,40 @@ func (u *ArtifactBucketUploader) uploadAssets(assets []asset) error {
+	hash := sha256.New()
+	for _, asset := range assets {
+		// hash.Write is documented to never return an error
+		hash.Write(asset.content)
```
If `asset` has changed its path that doesn't result in an reorder, do we want to produce a different hash?

dannyrandall replied:
good catch!! i've included the mapping file content in the hash which should cover this case.

iamhopaul123 commented on internal/pkg/cli/deploy/static_site.go:
```diff
@@ -43,16 +41,14 @@ func NewStaticSiteDeployer(in *WorkloadDeployerInput) (*staticSiteDeployer, erro
-			PathPrefix:       "local-assets",
-			AssetMappingPath: assetMappingPath,
+			FS:                  svcDeployer.fs,
+			AssetDir:            "local-assets",
```
Should we make "local-assets" a constant so that it can be reused for `AssetMappingFileDir`?

iamhopaul123 commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -140,30 +143,40 @@ func (u *ArtifactBucketUploader) uploadAssets(assets []asset) error {
+	// hash using the sorted list so order-only changes in assets (caused by
+	// manifest or walkFn reordering) doesn't result in a mapping name change.
+	hash := sha256.New()
+	for _, asset := range assets {
```
Very nit just a note that i don't think `asset.localPath` matters

dannyrandall replied:
chatted offline, we are only hashing content

iamhopaul123 commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -140,30 +143,40 @@ func (u *ArtifactBucketUploader) uploadAssets(assets []asset) error {
 
-	if err := u.Upload(u.AssetMappingPath, bytes.NewBuffer(data)); err != nil {
-		return fmt.Errorf("upload to %q: %w", u.AssetMappingPath, err)
+	path := path.Join(u.AssetMappingFileDir, hex.EncodeToString(hash.Sum(nil)))
```
can we avoid `path` as a var name? as it overrides the pkg name

iamhopaul123 reviewed (commented):
LGTM :shipit: only nits

dannyrandall committed:
addr fb from @Lou1415926:
- fix docstring
- hash mapping data so dst changes are included
- dedupe uploaded assets

dannyrandall committed:
fix/add tests for sorting/dedupe

dannyrandall committed:
fb from @iamhopaul123:
- don't shadow `path` import
- const for `local-assets` folder

Lou1415926 commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -174,9 +175,31 @@ func (u *ArtifactBucketUploader) uploadAssetMappingFile(assets []asset) (string,
-	if err := u.Upload(path, bytes.NewBuffer(data)); err != nil {
+	// include mapping file in hash so any src->dst
+	// mapping changes result in a new hash
+	hash.Write(data)
```
I'm trying to enumerate what would happen ⬇️ 💭 which looks good to me!!

Change content of a file ➡️  `ArtifactBucketPath` changes, i.e. the path to which it is uploaded is changed (because the path contains the hash) ➡️ `data` changes ➡️  This `hash` changes, because both `asset.content` and `data` change,

Change `dest` of an upload ➡️ `ServiceBucketPath` changes ➡️  This `hash` changes, because of `data`.

Move a file to a new local path (i.e. change `localPath` without changing `content`) ➡️ `localPath` ➡️  This `hash` does not change

dannyrandall replied:
based on this ^ i realized we can just hash the asset mapping file and not bother hashing the actual asset file content, since the `ArtifactBucketPath` will only change if the file content changes. updated the pr!

dannyrandall committed:
only hash the mapping file!

mergify[bot] committed:
Merge branch 'mainline' into chore/mapping-hash

efekarakus commented on internal/pkg/deploy/upload/asset/asset.go:
```diff
@@ -140,16 +143,20 @@ func (u *ArtifactBucketUploader) uploadAssets(assets []asset) error {
 
-// uploadAssetMappingFile uploads a JSON file to u.AssetMappingPath containing
-// the current location of each file in the artifact bucket and the desired location
+// uploadAssetMappingFile uploads a JSON file containing the current
```
nit: why the "current" distinction, can the objects move in the artifact bucket?

dannyrandall replied:
ah i see that might be a little confusing, i could probably just drop the word. i meant "location of the file _right now_" as opposed to the service bucket location

efekarakus reviewed (commented):
Looks great!

The PR description says:
>By making the file name based on a hash of the content, **the state machine will only be triggered if any of the uploaded content changes.**

How do you achieve that in the state machine?

dannyrandall commented:
> How do you achieve that in the state machine?

good question! the mapping file path [is an input](https://github.com/aws/copilot-cli/blob/222ea9cf4fab097abd7229380d661ba1a5f66fae/internal/pkg/template/templates/workloads/services/static-site/cf.yml#LL117C32-L117C32) to the custom resource that triggers the state machine - so cloudformation will only call `UPDATE` on the state-machine-trigger-lambda if the file path changes 😊
